File: app/src/main/java/com/sdex/activityrunner/intent/history/HistoryListAdapter.java
package com.sdex.activityrunner.intent.history;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.ContextMenu;
import android.view.ContextMenu.ContextMenuInfo;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import com.sdex.activityrunner.R;
import com.sdex.activityrunner.db.history.HistoryModel;
import com.sdex.activityrunner.intent.dialog.source.ActionSource;
import com.sdex.activityrunner.intent.dialog.source.MimeTypeSource;
import com.sdex.activityrunner.intent.param.None;
import java.util.ArrayList;
import java.util.List;

public class HistoryListAdapter extends
  RecyclerView.Adapter<HistoryListAdapter.ViewHolder> {

  public static final int MENU_ITEM_REMOVE = 0;

  private final Callback callback;
  private List<HistoryModel> items = new ArrayList<>();
  private int contextMenuItemPosition;

  public HistoryListAdapter(HistoryListAdapter.Callback callback) {
    this.callback = callback;
  }

  @NonNull
  @Override
  public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent,
    int viewType) {
    final LayoutInflater inflater = LayoutInflater.from(parent.getContext());
    final View view = inflater.inflate(R.layout.item_history, parent, false);
    return new ViewHolder(view);
  }

  @Override
  public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
    final HistoryModel item = items.get(position);
    holder.bind(item, callback);
    holder.itemView.setOnLongClickListener(v -> {
      setContextMenuItemPosition(holder.getAdapterPosition());
      return false;
    });
  }

  @Override
  public int getItemCount() {
    return items.size();
  }

  @Override
  public long getItemId(int position) {
    return items.get(position).getId();
  }

  @Override
  public void onViewRecycled(@NonNull ViewHolder holder) {
    holder.itemView.setOnLongClickListener(null);
    super.onViewRecycled(holder);
  }

  public HistoryModel getItem(int position) {
    return items.get(position);
  }

  public int getContextMenuItemPosition() {
    return contextMenuItemPosition;
  }

  public void setContextMenuItemPosition(int contextMenuItemPosition) {
    this.contextMenuItemPosition = contextMenuItemPosition;
  }

  public void setItems(List<HistoryModel> items) {
    this.items = items;
    notifyDataSetChanged();
  }

  public interface Callback {

    void onItemClicked(HistoryModel item, int position);
  }

  public static class ViewHolder extends RecyclerView.ViewHolder
    implements View.OnCreateContextMenuListener {

    private final ActionSource actionSource = new ActionSource();
    private final MimeTypeSource mimeTypeSource = new MimeTypeSource();

    public final TextView packageName;
    public final TextView className;
    public final TextView action;
    public final TextView data;
    public final TextView mimeType;
    public final TextView extras;
    public final TextView categories;
    public final TextView flags;

    public ViewHolder(View itemView) {
      super(itemView);
      this.packageName = itemView.findViewById(R.id.package_name);
      this.className = itemView.findViewById(R.id.class_name);
      this.action = itemView.findViewById(R.id.action);
      this.data = itemView.findViewById(R.id.data);
      this.mimeType = itemView.findViewById(R.id.mime_type);
      this.extras = itemView.findViewById(R.id.extras);
      this.categories = itemView.findViewById(R.id.categories);
      this.flags = itemView.findViewById(R.id.flags);
    }

    public void bind(HistoryModel item, Callback callback) {
      packageName.setText(getValueOrPlaceholder(item.getPackageName()));
      className.setText(getValueOrPlaceholder(item.getClassName()));
      action.setText(actionSource.getItem(item.getAction()));
      data.setText(getValueOrPlaceholder(item.getData()));
      mimeType.setText(mimeTypeSource.getItem(item.getMimeType()));
      extras.setText(isNotEmpty(item.getExtras()));
      categories.setText(isNotEmpty(item.getCategories()));
      flags.setText(isNotEmpty(item.getFlags()));

      itemView.setOnClickListener(v -> callback.onItemClicked(item, getAdapterPosition()));
      itemView.setOnCreateContextMenuListener(this);
    }

    private int isNotEmpty(String value) {
      return TextUtils.isEmpty(value) ? R.string.no : R.string.yes;
    }

    private String getValueOrPlaceholder(String value) {
      if (TextUtils.isEmpty(value)) {
        return None.VALUE;
      }
      return value;
    }

    @Override
    public void onCreateContextMenu(ContextMenu menu, View v, ContextMenuInfo menuInfo) {
      menu.setHeaderTitle(R.string.history_item_dialog_title);
      menu.add(Menu.NONE, MENU_ITEM_REMOVE, Menu.NONE, R.string.history_item_dialog_remove);
    }
  }
}

File: app/src/main/java/com/sdex/activityrunner/intent/history/HistoryListAdapter.kt
package com.sdex.activityrunner.intent.history

import android.view.ContextMenu
import android.view.ContextMenu.ContextMenuInfo
import android.view.LayoutInflater
import android.view.Menu
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.sdex.activityrunner.R
import com.sdex.activityrunner.databinding.ItemHistoryBinding
import com.sdex.activityrunner.db.history.HistoryModel
import com.sdex.activityrunner.intent.param.None

class HistoryListAdapter(
    private val callback: Callback
) : ListAdapter<HistoryModel, HistoryListAdapter.ViewHolder>(DIFF_CALLBACK) {

    var contextMenuItemPosition: Int = 0

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        return ViewHolder(ItemHistoryBinding.inflate(inflater, parent, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(getItem(position), callback)
        holder.itemView.setOnLongClickListener {
            contextMenuItemPosition = holder.bindingAdapterPosition
            false
        }
    }

    override fun getItemId(position: Int): Long =
        getItem(position)?.id?.toLong() ?: 0L

    override fun onViewRecycled(holder: ViewHolder) {
        holder.itemView.setOnLongClickListener(null)
        super.onViewRecycled(holder)
    }

    interface Callback {

        fun onItemClicked(item: HistoryModel, position: Int)
    }

    class ViewHolder(
        private val binding: ItemHistoryBinding
    ) : RecyclerView.ViewHolder(binding.root),
        View.OnCreateContextMenuListener {

        fun bind(item: HistoryModel, callback: Callback) {
            binding.packageName.text = getValueOrPlaceholder(item.packageName)
            binding.className.text = getValueOrPlaceholder(item.className)
            binding.action.text = getValueOrPlaceholder(item.action)
            binding.data.text = getValueOrPlaceholder(item.data)
            binding.mimeType.text = getValueOrPlaceholder(item.mimeType)
            binding.extras.setText(isNotEmpty(item.extras))
            binding.categories.setText(isNotEmpty(item.categories))
            binding.flags.setText(isNotEmpty(item.flags))

            binding.root.setOnClickListener {
                callback.onItemClicked(item, bindingAdapterPosition)
            }
            binding.root.setOnCreateContextMenuListener(this)
        }

        private fun isNotEmpty(value: String?): Int {
            return if (value.isNullOrEmpty()) R.string.no else R.string.yes
        }

        private fun getValueOrPlaceholder(value: String?): String {
            return if (value.isNullOrEmpty()) None.VALUE else value
        }

        override fun onCreateContextMenu(menu: ContextMenu, v: View, menuInfo: ContextMenuInfo?) {
            menu.setHeaderTitle(R.string.history_item_dialog_title)
            menu.add(
                Menu.NONE, MENU_ITEM_ADD_SHORTCUT, Menu.NONE,
                R.string.history_item_dialog_add_shortcut
            )
            menu.add(
                Menu.NONE, MENU_ITEM_EXPORT_URI, Menu.NONE,
                R.string.history_item_dialog_export_uri
            )
            menu.add(
                Menu.NONE, MENU_ITEM_REMOVE, Menu.NONE,
                R.string.history_item_dialog_remove
            )
        }
    }

    companion object {

        const val MENU_ITEM_REMOVE = 0
        const val MENU_ITEM_ADD_SHORTCUT = 1
        const val MENU_ITEM_EXPORT_URI = 2

        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<HistoryModel>() {

            override fun areItemsTheSame(oldItem: HistoryModel, newItem: HistoryModel) =
                oldItem.id == newItem.id

            override fun areContentsTheSame(oldItem: HistoryModel, newItem: HistoryModel) =
                oldItem == newItem
        }
    }
}
